### ISSUE TITLE
Add support for headless OpenGL

### ISSUE THREAD

**Issue by nsthorat** [tensorflow/tfjs-core#234]
deeplearn.js can be used in a headless nodejs environment with this tool:
https://github.com/stackgl/headless-gl

Floating point textures are not supported in headless-gl, but we have a switch that can store floats as 4 unsigned bytes using fixed precision floats. This already may work.

The relevant flag is `WEBGL_FLOAT_TEXTURE_ENABLED`, which is controlled by this function:
https://github.com/PAIR-code/deeplearnjs/blob/master/src/environment.ts#L99

We will have to detect when we are in the nodejs environment and make sure this returns false (it may already return false by running this test).

headless-gl will return a new `WebGLRenderingContext`, this can be passed directly to the constructor of `NDArrayMathGPU` to be used in this context.

We don't want to add headless-gl as a real dependencies, so for an example script, we should have people manually run `npm install headless-gl`. If we can show a simple example of this in our repository, that would be phenomenal.

We should then be able to support headless environments!

**Comment by arthurjdam** [comment]
Side-note; referencing #158, I'm working on (fully GPU accelerated) headless support via OpenCL (which removes the need for a fb/xvfb, I guess). Not exactly what you're mentioning here but it should serve the same purpose eventually.

Making some (rather slow, but steady) progress in here: [arthurjdam/deeplearnjs/feature-opencl](https://github.com/arthurjdam/deeplearnjs/tree/feature-opencl).

**Comment by nsthorat** [comment]
Yup, similar, though I think headless OpenGL we may be able do without many changes. OpenCL, in theory, should be much faster, in the headless OpenGL we're 1) using GPGPU shaders which is going to be slower than CL and 2) we have to use 4 byte integer quantization so we lose a little precision and we're slower than working directly with 32 bit floats.

**Comment by nsthorat** [comment]
How is this going Arthur?

**Comment by arthurjdam** [comment]
Progress is not particularly fast, but it's going! Matmul and unary ops are working, though it's taking some time to figure out how to piece it together so it's relatively aligned with WebGL's interface. I'm structuring some things out, but any insights on how to nicely perform initialization and batching would be super-helpful!

Using OpenCL 1.2 (since that just happens to be what my Macbook supports), performance tests look pretty solid, and allow for delegation to multiple cpu's/gpu's, as long as I don't try to make anything return doubles (so it's 32-bit floats for everyone) I'm getting quite good behavior from everything.

**Comment by arthurjdam** [comment]
Out of sheer laziness, I've be playing with adding node-webgl (webgl-type bindings around opengl) which seems to be relatively straight-forward. I've been running into an error that I could use a hand on if anyone's more knowledgeable: https://github.com/arthurjdam/deeplearnjs/tree/feature-headless-gl

The current build will throw an error when binding vertexAttrib buffers, but I'm not yet sure why. Commenting out `bindVertexBufferToProgramAttribute` for `clipSpacePos` will make it -at least not throw errors- but obviously breaks the engine. Any ideas?

**Comment by dsmilkov** [comment]
Unfortunately, I'm not a WebGL/OpenGL expert, but maybe [this page](https://www.khronos.org/webgl/wiki_1_15/index.php/WebGL_and_OpenGL_Differences#Vertex_Attribute_0) talking about the differences between WebGL and OpenGL might help.

Also might be good to ask the node-webgl author @mikeseven for ideas.

**Comment by arthurjdam** [comment]
Yeah I _think_ I've processed that doc on the web/open-gl differences. Fingers crossed @mikeseven can shed some light on the situation!

**Comment by mikeseven** [comment]
Sure, what is the question exactly?

--mike

**Comment by arthurjdam** [comment]
@mikeseven I sent an issue in the node-webgl repo with some explanation https://github.com/mikeseven/node-webgl/issues/59

**Comment by mikeseven** [comment]
I believe the error is due to changes in nodejs internals. I need to upgrade my packages to the new nodejs/v8 native interfaces.

--mike

________________________________

**Comment by arthurjdam** [comment]
Thanks for looking into it @mikeseven! Let me know if I can do anything to help
By the by; not sure if this helps, but, the demos (i.e. the test/lesson0*.js's) work just fine, so not sure what the discrepancy is. Tested under a few different Node 8 versions, though nothing lower.

**Comment by mikeseven** [comment]
The first lessons are very simple use of GL API. This shouldn't have changed but the other samples especially using vertex buffers and async call backs have changed because v8 has changed significantly.

At the same time, nodejs has introduced n-api recently and it's better than the previous native API.

Since n-api is the future, I need to rewrite these parts for node-webgl and node-webcl and test if performance and async call backs are still behaving the same.

I'm not sure when I will have time as I'm pretty full until end of November but when I have some free time, I'll start ;-)

--mike

________________________________

**Comment by erikwilson** [comment]
I found that if I can use NodeJs with DeepLearn if I added a shim for location:
```
global.window = {
  location: {
    search: ""
  }
}
```
and used NDArrayMathCPU instead of NDArrayMathGPU for doing math. In fact for small graphs using the CPU can be 10x faster than using the GPU, is there a general matrix size where using the GPU becomes more efficient?

I played with using headless-gl and node-webgl modules a little but they seemed to be lacking enough WebGL extensions not to be worthwhile.

Also the location search dependency should probably be removed...
